pkg/mcs/tso/server/server.go

```diff
@@ -16,77 +16,106 @@ package server
 
 import (
 	"context"
+	"crypto/tls"
+	"fmt"
+	"math/rand"
+	"net"
 	"net/http"
+	"net/url"
 	"os"
 	"os/signal"
+	"path"
+	"strconv"
+	"strings"
 	"sync"
+	"sync/atomic"
 	"syscall"
 	"time"
 
 	grpcprometheus "github.com/grpc-ecosystem/go-grpc-prometheus"
 	"github.com/pingcap/kvproto/pkg/diagnosticspb"
 	"github.com/pingcap/kvproto/pkg/tsopb"
 	"github.com/pingcap/log"
+	"github.com/pingcap/sysutil"
+	"github.com/soheilhy/cmux"
 	"github.com/spf13/cobra"
 	bs "github.com/tikv/pd/pkg/basicserver"
 	"github.com/tikv/pd/pkg/errs"
+	"github.com/tikv/pd/pkg/member"
+	"github.com/tikv/pd/pkg/systimemon"
 	"github.com/tikv/pd/pkg/tso"
+	"github.com/tikv/pd/pkg/utils/etcdutil"
 	"github.com/tikv/pd/pkg/utils/grpcutil"
 	"github.com/tikv/pd/pkg/utils/logutil"
 	"github.com/tikv/pd/pkg/utils/metricutil"
 	"github.com/tikv/pd/pkg/utils/tsoutil"
 	"github.com/tikv/pd/pkg/versioninfo"
 	"go.etcd.io/etcd/clientv3"
+	"go.etcd.io/etcd/pkg/types"
 	"go.uber.org/zap"
 	"google.golang.org/grpc"
 	"google.golang.org/grpc/codes"
 	"google.golang.org/grpc/status"
 )
 
-// If server doesn't implement all methods of bs.Server, this line will result in a clear
-// error message like "*Server does not implement bs.Server (missing Method method)"
+const (
+	leaderTickInterval = 50 * time.Millisecond
+	// tsoRootPath for all tso servers.
+	tsoRootPath      = "/tso"
+	tsoClusterIDPath = "/tso/cluster_id"
+	// defaultGRPCGracefulStopTimeout is the default timeout to wait for grpc server to gracefully stop
+	defaultGRPCGracefulStopTimeout = 5 * time.Second
+	// defaultHTTPGracefulShutdownTimeout is the default timeout to wait for http server to gracefully shutdown
+	defaultHTTPGracefulShutdownTimeout = 5 * time.Second
+)
+
 var _ bs.Server = (*Server)(nil)
 
-// Server is the TSO server, and it implements bs.Server and tso.GrpcServer.
+// Server is the TSO server, and it implements bs.Server.
 type Server struct {
 	diagnosticspb.DiagnosticsServer
 
+	// Server state. 0 is not serving, 1 is serving.
+	isServing int64
 	// Server start timestamp
 	startTimestamp int64
 
-	ctx       context.Context
-	name      string
-	clusterID uint64
+	ctx              context.Context
+	serverLoopCtx    context.Context
+	serverLoopCancel func()
+	serverLoopWg     sync.WaitGroup
+
+	cfg         *tso.Config
+	name        string
+	clusterID   uint64
+	rootPath    string
+	backendUrls []url.URL
+
+	// for the primary election in the TSO cluster
+	participant *member.Participant
```

Comment:
Same question as above.

Reply:
Same answer as above. The other components might call TSO Server.Member()/GetLeader() to get the underlying member struct which is pdpb.Member for now and will be changed to different type like tsopb.Member in the current pr or the next pr. We mainly use the interface in the underlying AllocatorManager and Local/Global TSO Allocator.

Reply:
I'll move the Member interface into pkg/tso

Reply:
done